src/sys/DISK.ASM

```diff
@@ -1007,32 +1007,70 @@ fat_entry:                                      ; Read/write a specific entry in
+    jb .address                                 ; If not, proceed
 
     clc                                         ; Otherwise, auto-align
     call align_fat                              ; Perform an alignment
     jc .abort                                   ; If error, abort
     jmp .check_entry                            ; And repeat
 
-.ins:
+.address:
     push 0x3000                                 ; The segment where the FAT is loaded
     pop es                                      ; Work from here
 
     shl si, 0x1                                 ; Two bytes per entry
+    mov bx, 0x0000                              ; Empty bitmask
+
+    cmp word [gs:cluster_max], 0x0FF7           ; Is this FAT12?
+    jae .ins                                    ; If not, proceed
+
+    mov ax, 0x0003                              ; Otherwise, 3 nibbles per entry
+    mul si                                      ; Multiply by 3
+
+    shr ax, 0x2                                 ; Remove the FAT16 remnant, and convert nibbles to bytes
+    mov bx, 0xF000                              ; Mask out the final three nibbles (12 bits)
+
+    test si, 0x0002                             ; Is this entry even? (We shifted left)
+    mov si, ax                                  ; The product is our address
+    jz .ins                                     ; If so, proceed
 
+    shl dx, 0x4                                 ; Otherwise, use left three nibbles
+    mov bx, 0x000F                              ; Reflect this in our bitmask
+
+.ins:
     popf                                        ; Restore flags
-    jnc .done                                   ; If CF cleared, proceed to the end
+    jnc .read                                   ; If read, proceed to read the entry
+
+    and word [es:si], bx                        ; Otherwise, clear out the bytes we're writing
+    not bx                                      ; Invert mask
+    and dx, bx                                  ; Clear out the bytes we're not writing
 
-    mov word [es:si], dx                        ; Otherwise, overwrite this entry
+    or word [es:si], dx                         ; Overwrite this entry
     clc                                         ; No error
 
-.done:
+    jmp .done                                   ; We're done!
+
+.read:
     mov dx, word [es:si]                        ; Read the corresponding entry in the FAT (or what we just wrote)
+    not bx                                      ; Invert mask
+    and dx, bx                                  ; Clear the bytes that aren't part of the entry
+
+    cmp bx, 0xFFF0                              ; Is the entry to the left?
+    clc                                         ; No error
+    jne .done                                   ; If not, we're done!
+
+    shr dx, 0x4                                 ; Otherwise, shift bits into place
+    clc                                         ; No error
+
+.done:
```

Comment:
Now using a bitmask to determine which bits are part of the entry we're interested in; it is used to clear out the proper bytes, so that OR can be used to write an entry. Implemented logic to handle FAT12 and its entry addressing.